### PR TITLE
Astar

### PR DESCRIPTION
**Pull request by wyxzou**
Add A* algorithm.

### CODE REVIEW

File: code/controller/astar.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    }
+    if (isValid(cur_x, cur_y+1, maxRow, maxCol)) {
+        neighbours.push_back(Astar::Coord{cur_x, cur_y+1});
+    }
+
+    return neighbours;
+}
+
+// For testing purposes
+Astar::grid Astar::makeTerrain() {
```
Comment (Mogball) [inline code/controller/astar.cpp]: add this to `test/controller/astar_test.cpp` and see other files for testing examples

File: code/controller/astar.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    while (cur != start) {
+        path.push_back(cur);
+        cur = parent[cur];
+    }
+    path.push_back(start);
+}
+
+void Astar::searchPath(Astar::grid terrain, Astar::Coord start, Astar::Coord dest) {
+    // idea for using maps to keep track of path information came from redblobgames.com
+    map<Astar::Coord, Astar::Coord> parent;
```
Comment (Mogball) [inline code/controller/astar.cpp]: technically if you used `unordered_map` and made a hash function for `Astar::Coord` you could design a perfect hash function for a given hash table size so you'd never get collisions, then it'd be way faster than using `map` which has an RB tree behind it

Comment (Mogball) [reply code/controller/astar.cpp]: actually that's probably more effort than it's worth

Comment (Mogball) [inline code/controller/astar.cpp]: couldn't you get away with just a list

File: code/controller/astar.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    bool isValid(int, int, int, int);
+
+    std::vector<Coord> getNeighbours(Coord, int, int);
+    grid makeTerrain();
+
+    void backtrack(Coord, Coord, std::map<Coord, Coord>);
+
+public:
+    Astar();
+    void searchPath(grid terrain, Coord cur, Coord dest);
```
Comment (Mogball) [inline code/controller/astar.h]: how do you get the path after it's calculateddddddddddddddddddddddddddddddddd

Comment (wyxzou) [reply code/controller/astar.h]: backtrack adds it to the vector path (I'll change it to a stack)

**Comment by Mogball** [comment]
@wyxzou  looks like a test is failing

**Pull request merged**